syntax = "proto3";

package messages;

option go_package = "main/pkg/messages";

// ===== TRANSCRIPTION TASK =====
message MessageTranscriptionTask {
  string file_url = 1;
  double start_time = 2;
  double end_time = 3;
}

// ===== CONVERT TASK =====
message MessageConvertTask {
  string file_url = 1;
}

// ===== DIARIZE TASK =====
message MessageDiarizeTask {
  string converted_file_url = 1;
}

// ===== SEGMENT STRUCTURE =====
message Segment {
  int32 speaker = 1;
  double start_time = 2;
  double end_time = 3;
}

message ConvertTaskResponse {
  string converted_file_url = 1;
  double audio_len = 2;
}

message DiarizeTaskResponse {
  int32 num_of_speakers = 1;
  repeated Segment segments = 2;
}

message TranscriptionTaskResponse {
  string transcription = 1;
}

message ErrorTaskResponse {
  string error = 1;
}

// ===== WRAPPER TASK =====

message WrapperTask {
  string task_id = 1;
  oneof task {
    MessageConvertTask convert = 2;
    MessageDiarizeTask diarize = 3;
    MessageTranscriptionTask transcription = 4;
  }
}

message WrapperResponse {
  string task_id = 1;
  oneof payload {
    ConvertTaskResponse convert = 2;
    DiarizeTaskResponse diarize = 3;
    TranscriptionTaskResponse transcription = 4;
    ErrorTaskResponse error = 5;
  }
}